File: go/1652.go
package main

func sum(code []int, i, j int) int {
	sum := 0
	d := j - i
	i %= len(code)
	if i < 0 {
		i += len(code)
	}
	j = i + d
	for ; i <= j; i++ {
		sum += code[i%len(code)]
	}
	return sum
}

func decrypt(code []int, k int) []int {

	res := []int{}
	for i := 0; i < len(code); i++ {
		if k > 0 {
			res = append(res, sum(code, i+1, i+k))
		} else {
			res = append(res, sum(code, i+k, i-1))
		}
	}
	return res
}

File: js/1652.js
const assert = require("assert");

var sum = function(code, i, j) {
    // console.log({ i, j });
    var diff = j - i
    i = i % code.length
    if (i < 0) {
        i += code.length
    }
    j = i + diff
    for(var s = 0, k = i; k <= j; k++) {
        s += code[k % code.length]
    }
    return s
}

/**
 * @param {number[]} code
 * @param {number} k
 * @return {number[]}
 */
 var decrypt = function(code, k) {

    // console.log('decrypt');
    return code.map((c, i, code) => {
        if(k > 0) {
            return sum(code, i + 1, i + k)
        } else {   
            return sum(code, i + k, i - 1)
        }
    })
};

var code = [5,7,1,4], k = 3
var res = [12,10,16,13]
assert.deepEqual(decrypt(code, k), res)
code = [1,2,3,4], k = 0
res = [0,0,0,0]
assert.deepEqual(decrypt(code, k), res)
code = [2,4,9,3], k = -2
res = [12,5,6,13]
assert.deepEqual(decrypt(code, k), res)